data_model/primitives/src/conststr.rs

```diff
@@ -0,0 +1,650 @@
+            inlined: InlinedString::new(),
+        }
+    }
+
+    #[inline]
+    const fn is_inlined(&self) -> bool {
+        // SAFETY: access to the MSB is always safe regardless of the correct variant.
+        self.inlined().is_inlined()
+    }
+
+    #[allow(unsafe_code)]
+    #[inline]
+    const fn inlined(&self) -> &InlinedString {
+        // SAFETY: safe to access if `is_inlined` == `true`.
+        unsafe { &self.inlined }
+    }
+
+    #[allow(unsafe_code)]
+    #[inline]
+    const fn boxed(&self) -> &ManuallyDrop<BoxedString> {
+        // SAFETY: safe to access if `is_inlined` == `false`.
+        unsafe { &self.boxed }
+    }
+}
+
+impl<T: ?Sized> AsRef<T> for ConstString
+where
+    InlinedString: AsRef<T>,
+    BoxedString: AsRef<T>,
+{
+    #[inline]
+    fn as_ref(&self) -> &T {
+        if self.is_inlined() {
+            self.inlined().as_ref()
+        } else {
+            self.boxed().as_ref()
+        }
+    }
+}
+
+impl Deref for ConstString {
+    type Target = str;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.as_ref()
+    }
+}
+
+impl Hash for ConstString {
+    #[inline]
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        (**self).hash(state)
+    }
+}
+
+impl Ord for ConstString {
+    #[inline]
+    fn cmp(&self, other: &Self) -> Ordering {
+        Ord::cmp(&**self, &**other)
+    }
+}
+
+/// Can't be derived.
+impl PartialOrd for ConstString {
+    #[inline]
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for ConstString {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        PartialEq::eq(&**self, &**other)
+    }
+}
+
+macro_rules! impl_eq {
+    ($($ty:ty,)*) => {
+        impl_eq!($($ty),*);
+    };
+    ($($ty:ty),*) => {$(
+        impl PartialEq<$ty> for ConstString {
+            // Not possible to write macro uniformly for different types otherwise.
+            #[allow(clippy::string_slice, clippy::deref_by_slicing)]
+            #[inline]
+            fn eq(&self, other: &$ty) -> bool {
+                PartialEq::eq(&self[..], &other[..])
+            }
+        }
+
+        impl PartialEq<ConstString> for $ty {
+            // Not possible to write macro uniformly for different types otherwise.
+            #[allow(clippy::string_slice, clippy::deref_by_slicing)]
+            #[inline]
+            fn eq(&self, other: &ConstString) -> bool {
+                PartialEq::eq(&self[..], &other[..])
+            }
+        }
+    )*};
+}
+
+impl_eq!(String, str, &str);
+
+/// Can't be derived.
+impl Eq for ConstString {}
+
+impl<T> From<T> for ConstString
+where
+    T: TryInto<InlinedString>,
+    <T as TryInto<InlinedString>>::Error: Into<BoxedString>,
+{
+    #[inline]
+    fn from(value: T) -> Self {
+        match value.try_into() {
+            Ok(inlined) => Self { inlined },
+            Err(value) => Self {
+                boxed: ManuallyDrop::new(value.into()),
```

Comment:
I'm sensing a potential leak here. You're giving out to the user of the library an object which they should manually drop. That is not a good api